target/linux/ath79/generic/base-files/etc/hotplug.d/firmware/10-ath9k-eeprom

```diff
@@ -39,6 +40,10 @@ case "$FIRMWARE" in
 		caldata_extract "art" 0x1000 0x440
 		ath9k_patch_mac $(mtd_get_mac_ascii u-boot-env ethaddr)
 		;;
+	mikrotik,rb-wapg-5hact2hnd)
+		mikrotik_caldata_extract "hard_config" 0x1000 0x440 $FIRMWARE
+		ath9k_patch_mac $(macaddr_add $(cat /sys/class/net/eth0/address) +2)
```

Comment:
Please do not derive the MAC address based on other interfaces.
Best would be to read directly from flash if the address is found anywhere. Otherwise, don't calculate based on the eth0 device address, but use the source of this address directly, e.g.
`ath9k_patch_mac $(macaddr_add $(mtd_get_mac_text hard_config 0x10) +2)`

Despite, I wonder why ath10k address is read from caldata and ath9k is not

Reply:
Fixed it.

The ath10k caldata contains the actual valid MAC, while the ath9k just contains a fake 11 22 33 11 11 11 one :disappointed:.

Unpatched:
```
root@OpenWrt:/# hexdump /lib/firmware/ath9k-eeprom-ahb-18100000.wmac.bin 
0000000 0202 1122 3311 1111 534e 3a37 3645 3830
0000010 3142 3345 4233 4300 0000 0000 0000 1f00
0000020 3302 0000 0000 0400 4400 4d04 0300 08ff
```

Patched:
```
root@OpenWrt:/# hexdump /lib/firmware/ath9k-eeprom-ahb-18100000.wmac.bin 
0000000 0202 000c 42c0 ffee 534e 3a37 3645 3830
0000010 3142 3345 4233 4300 0000 0000 0000 1f00
0000020 3302 0000 0000 0400 4400 4d04 0300 08ff
```

Reply:
Have you checked the device's address without patching (cat /sys/class/ieee80211/phyX/macaddress)? Sometimes ath9k seems to read from other locations, too.

Reply:
I f I remove the ath9k patching line, I get a random MAC every time:

```
root@OpenWrt:/# cat /sys/class/ieee80211/phy10/macaddress
06:10:56:7f:3c:68
root@OpenWrt:/# rmmod ath9k; rm /lib/firmware/ath9k-eeprom-ahb-18100000.wmac.bin
 ; modprobe ath9k
[ 1527.107196] ath9k: ath9k: Driver unloaded
[ 1527.174629] ath9k 18100000.wmac: Direct firmware load for ath9k-eeprom-ahb-18100000.wmac.bin failed with error -2
[ 1527.185092] ath9k 18100000.wmac: Falling back to syfs fallback for: ath9k-eeprom-ahb-18100000.wmac.bin
[ 1528.036906] ath: phy11: eeprom contains invalid mac address: 11:22:33:11:11:11
[ 1528.044306] ath: phy11: random mac address will be used: 0e:04:26:83:ac:0f
[ 1528.071169] ieee80211 phy11: Atheros AR9550 Rev:0 mem=0xb8100000, irq=13
root@OpenWrt:/# cat /sys/class/ieee80211/phy11/macaddress
0e:04:26:83:ac:0f
```

Reply:
Okay, thanks for testing. So if the correct address is not in art elsewhere, use the ath9k_patch_mac $(macaddr_add $(mtd_get_mac_text hard_config 0x10) +2) option.